Notification Deduplication and NotificationMenu Refactor

zakhap opened the pull request:
## Description

This PR began simply to address notification duplication in the Notification Dropdown menu but also includes a factoring out of the NotificationMenu from the Header and PseudoHeader into its own component.

- Notifications Dropdown Menu / `notifications_dropdown_menu.ts`.
  - Simple refactor to reusable component
  - Sorts Notifications into related arrays through a reducer back to arrays of related notifications.
    - This should be looked at closely***
    - In short, the reducer sorts each Notification by their Subscription property's _objectId_ (the thread/comment/community in which the notification is directly responding to).
  - Uses new notification row, `m(HeaderBatchNotificationRow)` which takes in an array of notifications, rather than just a single notification, and then shows the same single-notification row or a modified batched row that when clicked, marks all the notifications in associated array as read. 
      - This is how deduplicating was accomplished.

- Unread notifications now are highlighted with proper style as requested in #234 

### Related Github Issue:
- https://github.com/hicommonwealth/commonwealth-oss/issues/147
- #234 



## Motivation and Context
Too many notifications from a single subscription is annoying, batched notifications show engagement while minimizing inbox clutter.

## How Has This Been Tested?
Just UI tests (see the deduplication working) and checking that when a batched row is clicked, that the whole array of notifications is being updated in the server. 

## Have proper tags been added (for bug, enhancement, breaking change)?
- [x] yes

raykyri commented:
Going to let Dillon take a look at this first, and then I can review afterwards

drewstone commented on client/scripts/views/components/header.ts:
```diff
@@ -35,28 +32,8 @@ const Header: m.Component<{}> = {
-        ]),
-      }),
+      app.isLoggedIn()
+        && m(NotificationsDrowdownMenu),
```
can be on same line

zakhap replied:
👍

drewstone commented on client/scripts/views/components/notifications_dropdown_menu.ts:
```diff
@@ -0,0 +1,52 @@
+const NotificationsDrowdownMenu: m.Component<{},{}> = {
+  view: (vnode) => {
+    const notifications = app.login.notifications
+      ? app.login.notifications.notifications.sort((a, b) => b.createdAt.unix() - a.createdAt.unix()) : [];
```
can separate the ":" onto a new line.

zakhap replied:
👍

zakhap commented:
> How are we currently testing the deduplication work? Do we have unit tests?

No unit tests... it's just a UI thing, there are no new routes to test

drewstone commented:
@zakhap would be beneficial to merge in `raymond.construct` into this now that tests are set up again. As I also mentioned in Slack, tests of this type of logic would be super helpful. We shouldn't consider any limitations to the things we can test. Use helper functions given an array of notifications and test that you get proper deduplication results.

dillchen commented:
This worked for me!

I added some issues/enhancements separately. These probably would be good to get @raykyri's  feedback on before implementing anything. https://github.com/hicommonwealth/commonwealth-oss/issues/220

dillchen commented:
@zakhap what's the status of this now?

zakhap commented:
I added and modified a few tests to make sure batch editing an array of notifications or subscriptions via related routes worked as intended. I think this is ready for another (hopefully final?) review

drewstone commented:
It would be good to merge in master to get the full test suite. Let me know if you want me to handle that @zakhap

drewstone commented on test/unit/api/subscriptions.spec.ts:
```diff
@@ -110,6 +111,41 @@ describe('Subscriptions Tests', () => {
+
+    it('should pause and unpause an array of subscription', async () => {
+      const subscriptions = [];
+      const subscription1: NotificationSubscription = await modelUtils.createSubscription({
```
Can you use a for loop to make this more concise?

drewstone replied:
This can still be more concise @zakhap

drewstone commented on test/unit/api/subscriptions.spec.ts:
```diff
@@ -155,6 +192,14 @@ describe('Subscriptions Tests', () => {
         body: 'hi you!',
         kind: 'forum',
       });
+      thread2 = await modelUtils.createThread({
```
what is this 2nd thread being used for?

zakhap replied:
to emit a 2nd notification for the routes to test an array > 1

drewstone replied:
where is that being tested? I'm not sure I'm seeing any check for that in the test itself? And an array > 1 of notifications? If that's what you're testing shouldn't there be an `expect(notificationsList.length).to.be.equal(2)`?

zakhap commented:
> Not sure I see any tests regarding deduplication. Similarly, since we're using the createSubscription route, we should also aim to get that to green coverage. Both create/delete subscriptions are now at yellow coverage.

Regarding deduplication, it's just a sorting/reducer function in the UI and then the routes handle an array of notification to mark-as-read. I emitted extra notifications to make sure the routes were handling an array of notifications and not just one.

zakhap commented:
@drewstone I wrote a bunch of tests, and we're in the 90th percentile for all the touched tests.

zakhap commented:
> Awesome getting more tests added. Can we also get checks for the error messages and consolidate them as I have been doing in previous tests? I think you reviewed examples in the past. We should aim to be as verbose as possible with these checks, so ensuring the right error message gets return is also crucial towards writing proper tests.

I ended up added Error imports for all tests that have error checks and reorganizing subscription related routes in the PR #215, please check it out over there! 

> where is that being tested? I'm not sure I'm seeing any check for that in the test itself? And an array > 1 of notifications? If that's what you're testing shouldn't there be an expect(notificationsList.length).to.be.equal(2)?

I ended up removing the 2nd createThread because `/viewNotification` was returning well more than 2 (18 exactly) notifications from all the other threads produced for other tests using the same jwt in the tests prior (but within `subscriptions.spec.ts` generally). 

> This can still be more concise @zakhap

I made this a loop!

zakhap commented:
Added styling to unread notifications in dropdown menu #234
